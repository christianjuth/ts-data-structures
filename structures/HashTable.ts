import LinkedList from './LinkedList';

interface HashTable<T> {
  columns: Array<LinkedList<T>>;
  maxLoad: number;
  mod: number;
}

class HashTable<T>{
  constructor(options?: { mod?: number, maxLoad?: number }) {
    let { maxLoad, mod } = Object.assign({
      maxLoad: 1,
      mod: 10
    }, options);
    this.mod = mod;
    this.maxLoad = maxLoad;
    this.columns = this.buildColumns();
  }

  private buildColumns(): Array<LinkedList<T>> {
    return new Array<LinkedList<T>>(this.mod);
  }

  private grow(min: number): void {
    console.log(min, this.mod + 1, Math.min(min, this.mod + 1));
    const oldColumns = this.columns;
    this.mod = Math.max(min, this.mod + 1);
    this.columns = this.buildColumns();
    oldColumns.forEach(col => {
      col.forEach(value => {
        this.add(value);
      });
    });
  }

  add(value: T): boolean {
    if(this.contains(value)) return false;
    let index = this.id(value) % this.mod;
    if(!this.columns[index]) {
      this.columns[index] = new LinkedList<T>();
    }
    if(this.columns[index].length() > (this.maxLoad - 1)) {
      this.grow(index);
      this.add(value);
    } else {
      this.columns[index].push(value);
    }
    return true;
  }

  remove(value: T): boolean {
    if(!this.contains(value)) return false;
    let index = this.id(value) % this.mod;
    return this.columns[index].remove(value);
  }

  contains(value: T): boolean {
    let index = this.id(value) % this.mod;
    if(!this.columns[index]) {
      return false;
    } else{
      return this.columns[index].contains(value);
    }
  }

  forEach(callback: (item: T) => void): void {
    this.columns.forEach(col => {
      col.forEach(callback);
    });
  }

  toString(): String {
    let output = '';
    this.columns.forEach((col, i) => {
      if(output.length > 0) output += '\n';
      output += `${i}) ${col.toString()}`;
    });
    return output;
  }

  traverse(): void {
    console.log(this.toString());
  }

  private stringId(str: string): number {
    var hash = 0, i, chr, len;
    for (i = 0, len = str.length; i < len; i++) {
      chr   = str.charCodeAt(i);
      hash  = ((hash << 5) - hash) + chr;
      hash |= 0; // Convert to 32bit integer
    }
    return hash;
  }

  private id(value: T): number {
    switch(typeof value) {
      case 'number':
        return value;
      case 'string':
        return this.stringId(value);
      default:
        throw new Error('failed to create id');
    }
  };
}

export default HashTable;