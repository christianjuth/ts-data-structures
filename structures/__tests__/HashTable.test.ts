import HashTable from '../HashTable';


describe("HashTable", () => {

  test("add", () => {
    let table = new HashTable<number>();
    table.add(5);
    table.add(6);
    expect(table.toString()).toMatchSnapshot();
  });

  test("double add", () => {
    let table = new HashTable<String>();
    expect(table.add('bob')).toBe(true);
    expect(table.add('bob')).toBe(false);
    expect(table.toString()).toMatchSnapshot();
  });

  test("contains", () => {
    let table = new HashTable<String>();
    table.add('bob');
    expect(table.contains('bob')).toBe(true);
    expect(table.contains('mike')).toBe(false);
  });

  test("remove", () => {
    let table = new HashTable<number>();
    table.add(5);
    table.add(8);
    expect(table.toString()).toMatchSnapshot();
    expect(table.remove(10)).toBe(false);
    expect(table.remove(5)).toBe(true);
    expect(table.toString()).toMatchSnapshot();
  });

  test("forEach", () => {
    let list = new HashTable<number>({
      mod: 4
    });
    // use small number so the
    // order of forEach is predictable
    list.add(0);
    list.add(1);
    list.add(2);
    list.add(3);
    let baseline: number[] = [0, 1, 2, 3],
      i = 0;
    list.forEach(item => {
      expect(item).toBe(baseline[i]);
      i++;
    });
  });

  test("isEmpty", () => {
    let list = new HashTable<number>();
    expect(list.isEmpty()).toBe(true);
    list.add(0);
    expect(list.isEmpty()).toBe(false);
  });

});