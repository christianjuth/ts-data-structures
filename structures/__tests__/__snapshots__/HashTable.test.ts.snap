// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HashTable add 1`] = `
"5) (5)
6) (6)"
`;

exports[`HashTable double add 1`] = `"7) (bob)"`;

exports[`HashTable remove 1`] = `
"5) (5)
8) (8)"
`;

exports[`HashTable remove 2`] = `
"5) ()
8) (8)"
`;
